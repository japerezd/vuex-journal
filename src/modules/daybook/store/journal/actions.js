// export const myAction = async ({commit}) => {}

import journalApi from '@/api/journalApi';

export const loadEntries = async ({ commit }) => {
  const { data } = await journalApi.get('/entries.json');
  if (!data) {
    return commit('setEntries', [])
  }
  const entries = [];
  for (let id of Object.keys(data)) {
    entries.push({
      id,
      ...data[id],
    });
  }
  commit('setEntries', entries);
};

export const updateEntry = async ({ commit }, entry) => {
    const id = entry.id;
    const dataToSave = {
      date: entry.date,
      text: entry.text,
      picture: entry.picture,
    }
    dataToSave.id = entry.id
    await journalApi.put(`/entries/${id}.json`, dataToSave);
    commit('updateEntry', {...dataToSave})
};

export const createEntry = async ({ commit }, entry) => {
  const dataToSave = {...entry}
  const { data } = await journalApi.post(`/entries.json`, dataToSave);
  entry.id = data.name
  commit('addEntry', entry)
  return data.name;
};

export const deleteEntry = async ({ commit }, id) => {
  await journalApi.delete(`/entries/${id}.json`);
  commit('deleteEntry',id )
};
