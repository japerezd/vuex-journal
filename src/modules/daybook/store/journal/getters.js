// export const myGetter = (state) => {
//     return state.somthing
// }

export const getEntriesByTerm = (state) => (term = '') => {
  if (term.length === 0) return state.entries;
  return state.entries.filter((entry) =>
    entry.text.toLowerCase().includes(term.toLocaleLowerCase())
  );
};

export const getEntryById = (state) => (id = '') => {
  const entry = state.entries.find(ent => ent.id === id);
  if (!entry) return;
  return { ...entry };
};
