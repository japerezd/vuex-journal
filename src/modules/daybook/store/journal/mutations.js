// export const myMutation = (state) => {

// }

export const setEntries = (state, entries) => {
  state.entries = [...state.entries, ...entries];
  state.isLoading = false;
};

export const updateEntry = (state, entry) => {
  const entryIndex = state.entries.findIndex((ent) => ent.id === entry.id);
  state.entries[entryIndex] = entry;
};

export const addEntry = (state, entry) => {
  state.entries = [entry, ...state.entries]
};

export const deleteEntry = (state, id) => {
    state.entries = state.entries.filter(entry => entry.id !== id);
};
