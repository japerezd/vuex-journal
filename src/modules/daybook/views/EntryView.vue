<template>
  <template v-if="entry">
    <div class="entry-title d-flex justify-content-between p-2">
      <div>
        <span class="text-success fs-3 fw-bold">{{ day }}</span>
        <span class="mx-1 fs-3">{{ month }}</span>
        <span class="mx-2 fs-4 fw-light">{{ yearDay }}</span>
      </div>

      <div>
        <input ref="imageSelector" hidden accept="image/png, image/jpeg" type="file" @change="onSelectedImage"/>
        <button v-if="entry.id" class="btn btn-danger mx-2" @click="onDeleteEntry">
          Borrar
          <i class="fa fa-trash-alt"></i>
        </button>
        <button class="btn btn-primary" @click="onSelectImage">
          Subir foto
          <i class="fa fa-upload"></i>
        </button>
      </div>
    </div>
    <hr />
    <div class="d-flex flex-column px-3 h-75">
      <textarea v-model="entry.text" placeholder="Que sucedio hoy?"></textarea>
    </div>
    <img
      v-if="entry.picture && !localImage"
      :src="entry.picture"
      alt="entry-picture"
      class="img-thumbnail"
    />
    <img
      v-if="localImage"
      :src="localImage"
      alt="entry-picture"
      class="img-thumbnail"
    />
  </template>

  <Fab :icon="'fa-save'" @on-click="saveEntry" />
</template>

<script>
import { defineAsyncComponent } from 'vue';
import { mapGetters, mapActions } from 'vuex';
import getDayMonthYear from '../helpers/getDayMonthYear';
import Swal from 'sweetalert2';
import uploadImage from '../helpers/uploadImage';

export default {
  name: 'EntryView',
  props: {
    id: {
      type: String,
      required: true,
    },
  },
  components: {
    Fab: defineAsyncComponent(() => import('../components/Fab.vue')),
  },

  data() {
    return {
      entry: null,
      localImage: null,
      file: null,
    };
  },

  watch: {
    id() {
      this.loadEntry();
    },
  },

  created() {
    this.loadEntry();
  },

  computed: {
    ...mapGetters('journal', ['getEntryById']),

    day() {
      const { day } = getDayMonthYear(this.entry.date);
      return day;
    },
    month() {
      const { month } = getDayMonthYear(this.entry.date);
      return month;
    },
    yearDay() {
      const { yearDay } = getDayMonthYear(this.entry.date);
      return yearDay;
    },
  },

  methods: {
    loadEntry() {
      let entry;
      if (this.id === 'new') {
        entry = {
          text: '',
          date: new Date().getTime(),
        };
      } else {
        entry = this.getEntryById(this.id);
        if (!entry) return this.$router.push({ name: 'no-entry' });
      }
      this.entry = entry;
    },
    async saveEntry() {
      Swal.fire({
        title: 'Espere por favor',
        allowOutsideClick: false
      })
      Swal.showLoading();
      const picture = await uploadImage(this.file);

      this.entry.picture = picture;

      if (this.entry.id) {
        // Actualizar
        await this.updateEntry(this.entry);
      } else {
        await this.createEntry(this.entry);
        this.$router.push({ name: 'entry', params: { id: this.entry.id } });
      }
      this.file = null;

      Swal.fire('Guardado', 'Entrada registrada con exito', 'success')
    },

    async onDeleteEntry() {
      const { isConfirmed } = await Swal.fire({
        title: 'Esta seguro?',
        text: 'Una vez borrado, no se puede recuperar',
        showDenyButton: true,
        confirmButtonText: 'Si, estoy seguro'
      })
      if (isConfirmed) {
        Swal.fire({
          title: 'Espere por favor',
          allowOutsideClick: false
        })
        Swal.showLoading();
        await this.deleteEntry(this.entry.id);
        this.$router.push({name: 'no-entry'})

        Swal.fire('Eliminado','', 'success')
      }
    },

    onSelectedImage(event) {
      const file = event.target.files[0]
      if (!file) {
        this.localImage = null;
        this.file = null;
        return;
      }
      this.file = file;
      const fr = new FileReader();
      fr.onload = () => this.localImage = fr.result
      fr.readAsDataURL(file)
    },

    onSelectImage() {
      this.$refs.imageSelector.click();
    },

    ...mapActions('journal', ['updateEntry', 'createEntry', 'deleteEntry']),
  },
};
</script>

<style lang="scss" scoped>
textarea {
  font-size: 20px;
  border: none;
  height: 100%;

  &:focus {
    outline: none;
  }
}

img {
  width: 200px;
  position: fixed;
  bottom: 150px;
  right: 20px;
  box-shadow: 0 5px 10px rgba($color: #000000, $alpha: 0.2);
}
</style>
