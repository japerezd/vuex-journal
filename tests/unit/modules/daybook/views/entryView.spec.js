import { shallowMount } from '@vue/test-utils';
import { createStore } from 'vuex';
import Swal from 'sweetalert2';
import journal from '@/modules/daybook/store/journal';
import { journalState } from '../../../mock-data/test-journal-state';

import EntryView from '@/modules/daybook/views/EntryView.vue';

const createVuexStore = (initialState) =>
  createStore({
    modules: {
      journal: {
        ...journal,
        state: { ...initialState },
      },
    },
  });

// utilizando las propiedades que utiliza el sweetalert2
jest.mock('sweetalert2', () => ({
  fire: jest.fn(),
  showLoading: jest.fn(),
  close: jest.fn(),
}));

describe('Pruebas en el EntryView component', () => {
  const store = createVuexStore(journalState);
  store.dispatch = jest.fn();

  const mockRouter = {
    push: jest.fn(),
  };

  let wrapper;

  beforeEach(() => {
    jest.clearAllMocks();
    wrapper = shallowMount(EntryView, {
      props: {
        id: '-MiJYhNIHihlBvVuDr_5',
      },
      global: {
        mocks: {
          $router: mockRouter,
        },

        plugins: [store],
      },
    });
  });

  test('debe de sacar al usuario porque el id no existe', () => {
    const wrapper = shallowMount(EntryView, {
      props: {
        id: 'Este id no existe en el store',
      },
      global: {
        mocks: {
          $router: mockRouter,
        },

        plugins: [store],
      },
    });

    expect(mockRouter.push).toHaveBeenCalledWith({ name: 'no-entry' });
  });

  test('debe de mostrar la entrada correctamente', () => {
    expect(wrapper.html()).toMatchSnapshot();
    expect(mockRouter.push).not.toHaveBeenCalledWith({ name: 'no-entry' });
  });

  test('debe de borrar la entrada y salir', async () => {
    Swal.fire.mockReturnValueOnce(Promise.resolve({ isConfirmed: true }));
    await wrapper.find('.btn-danger').trigger('click');
    expect(Swal.fire).toHaveBeenCalledWith({
      title: 'Esta seguro?',
      text: 'Una vez borrado, no se puede recuperar',
      showDenyButton: true,
      confirmButtonText: 'Si, estoy seguro',
    });

    // setTimeout(() => {
    expect(mockRouter.push).toHaveBeenCalled();
    expect(store.dispatch).toHaveBeenCalledWith('journal/deleteEntry', '-MiJYhNIHihlBvVuDr_5')
    //       done();
    // }, 1)
  });
});
