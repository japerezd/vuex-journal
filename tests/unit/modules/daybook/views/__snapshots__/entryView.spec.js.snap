// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el EntryView component debe de mostrar la entrada correctamente 1`] = `
<div class="entry-title d-flex justify-content-between p-2">
  <div><span class="text-success fs-3 fw-bold">29</span><span class="mx-1 fs-3">Agosto</span><span class="mx-2 fs-4 fw-light">2021, Domingo</span></div>
  <div><input hidden="" accept="image/png, image/jpeg" type="file"><button class="btn btn-danger mx-2"> Borrar <i class="fa fa-trash-alt"></i></button><button class="btn btn-primary"> Subir foto <i class="fa fa-upload"></i></button></div>
</div>
<hr>
<div class="d-flex flex-column px-3 h-75"><textarea placeholder="Que sucedio hoy?"></textarea></div><img src="https://res.cloudinary.com/dumihjct9/image/upload/v1606334667/zbk2wsw0tz3ozzb80wsg.jpg" alt="entry-picture" class="img-thumbnail">
<!--v-if-->
<fab-stub icon="fa-save"></fab-stub>
`;
